controllers/reconcilers/kserve_isvc_gateway_reconciler.go

```diff
@@ -0,0 +1,263 @@
+	secretHandler  resources.SecretHandler
+	gatewayHandler resources.GatewayHandler
+	deltaProcessor processors.DeltaProcessor
+}
+
+func NewKserveGatewayReconciler(client client.Client) *KserveGatewayReconciler {
+	return &KserveGatewayReconciler{
+		client:         client,
+		secretHandler:  resources.NewSecretHandler(client),
+		gatewayHandler: resources.NewGatewayHandler(client),
+		deltaProcessor: processors.NewDeltaProcessor(),
+	}
+}
+
+func (r *KserveGatewayReconciler) Reconcile(ctx context.Context, log logr.Logger, isvc *kservev1beta1.InferenceService) error {
+	log.V(1).Info("Reconciling KServe local gateway for Kserve InferenceService")
+
+	// return if URL is not set
+	if isvc.Status.URL == nil {
+		log.V(1).Info("Waiting for the URL as the Inference Service is not ready yet")
+		return nil
+	}
+
+	destinationSecretExist := true
+	if _, err := r.secretHandler.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: constants.IstioNamespace}); err != nil {
+		if errors.IsNotFound(err) {
+			destinationSecretExist = false
+		} else {
+			return err
+		}
+	}
+
+	if !destinationSecretExist {
+		secret := &corev1.Secret{}
+		for attempt := 1; attempt <= maxSecretCheckAttempts; attempt++ {
+			getSecretErr := r.client.Get(ctx, types.NamespacedName{Name: isvc.Name, Namespace: isvc.Namespace}, secret)
+			if getSecretErr == nil {
+				break
+			}
+			if errors.IsNotFound(getSecretErr) {
+				log.V(2).Info("The certificate secret is not created yet. Retrying...", "attempt_number", attempt)
+
+				time.Sleep(secretCheckInterval)
+			} else {
+				log.Error(getSecretErr, "Failed to retrieve the certificate secret for the InferenceService (ISVC)")
+				return getSecretErr
+			}
+		}
+		if err := r.copyServingCertSecretFromIsvcNamespace(ctx, secret); err != nil {
+			log.V(1).Error(err, "Failed to update KServe local gateway in the istio-system namespace")
+
+			return nil
+		}
+	}
+
+	// Create Desired resource
+	desiredResource, err := r.createDesiredResource(isvc)
+	if err != nil {
+		return err
+	}
+
+	// Get Existing resource
+	existingResource, err := r.getExistingResource(ctx)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Error(err, "Failed to find KServe local gateway in istio-system namespace")
+		}
+		return err
+	}
+
+	// Process Delta
+	if err = r.processDelta(ctx, log, desiredResource, existingResource); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *KserveGatewayReconciler) createDesiredResource(isvc *kservev1beta1.InferenceService) (*istioclientv1beta1.Gateway, error) {
+	hostname := fmt.Sprintf("%s.%s.%s", isvc.Name, isvc.Namespace, internalSerivceHostnameDomain)
+
+	desiredGateway := &istioclientv1beta1.Gateway{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "kserve-local-gateway",
+			Namespace: isvc.Namespace,
+		},
+		Spec: istiov1beta1.Gateway{
+			Servers: []*istiov1beta1.Server{
+				{
+					Hosts: []string{hostname},
+					Port: &istiov1beta1.Port{
+						Name:     isvc.Name,
+						Number:   8445,
```

Comment:
are these info expected to never change?

Reply:
yes, it is managed by odh-model-controller and if it is changed by something else, local communication will fail. So I don't think this information won't be changed.